Mathlib/Analysis/Normed/Affine/AsymptoticCone.lean

```diff
@@ -0,0 +1,428 @@
+  refine (Equiv.vaddConst p).iSup_congr fun _ => ?_
+  simp [add_vadd]
+
+theorem asymptoticNhds_eq_smul_vadd (v : V) (p : P) :
+    asymptoticNhds k P v = atTop (α := k) • 𝓝 v +ᵥ pure p := by
+  rw [← asymptoticNhds_eq_smul, asymptoticNhds_vadd_pure]
+
+omit [OrderTopology k] [IsStrictOrderedRing k] in
+theorem vadd_asymptoticNhds (u v : V) : u +ᵥ asymptoticNhds k P v = asymptoticNhds k P v := by
+  have ⟨p⟩ : Nonempty P := inferInstance
+  nth_rw 1 [← asymptoticNhds_vadd_pure v p]
+  simp_rw [← asymptoticNhds_vadd_pure v (u +ᵥ p), vadd_pure, ← Filter.map_vadd, map_map]
+  congr with v
+  exact vadd_comm u v p
+
+instance {v : V} : (asymptoticNhds k P v).NeBot := by
+  have ⟨p⟩ : Nonempty P := inferInstance
+  rw [asymptoticNhds_eq_smul_vadd v p]
+  infer_instance
+
+private theorem asymptoticNhds_zero' : asymptoticNhds k V (0 : V) = ⊤ := by
+  rw [← top_le_iff, top_eq_iSup_pure, iSup_le_iff]
+  intro v
+  rw [← map_const (f := atTop (α := k))]
+  have : (fun _ => v) =ᶠ[atTop (α := k)]
+      (Function.uncurry (· • ·)) ∘ (fun c => (c, c⁻¹ • v)) := by
+    filter_upwards [eventually_ne_atTop 0] with _ h
+    simp [h]
+  rw [map_congr this, ← map_map, asymptoticNhds_eq_smul, ← map₂_smul, ← map_prod_eq_map₂]
+  apply map_mono
+  have : Tendsto (fun c => (c, c⁻¹ • v)) (atTop (α := k)) _ :=
+    tendsto_id.prodMk <| tendsto_inv_atTop_zero.smul_const v
+  simpa
+
+@[simp]
+theorem asymptoticNhds_zero : asymptoticNhds k P (0 : V) = ⊤ := by
+  have ⟨p⟩ : Nonempty P := inferInstance
+  rw [← asymptoticNhds_vadd_pure 0 p, asymptoticNhds_zero', vadd_pure]
+  exact (Equiv.vaddConst p).surjective.filter_map_top
+
+theorem Filter.Tendsto.atTop_smul_nhds {α} {f : α → k} {g : α → V} {l : Filter α} {v : V}
+    (hf : Tendsto f l atTop) (hg : Tendsto g l (𝓝 v)) :
+    Tendsto (fun x => f x • g x) l (asymptoticNhds k V v) := by
+  rw [asymptoticNhds_eq_smul, ← map₂_smul, ← map_prod_eq_map₂]
+  exact tendsto_map.comp (hf.prodMk hg)
+
+theorem Filter.Tendsto.atTop_smul_const {α} {f : α → k} {l : Filter α} (v : V)
+    (hf : Tendsto f l atTop) :
+    Tendsto (fun x => f x • v) l (asymptoticNhds k V v) :=
+  hf.atTop_smul_nhds tendsto_const_nhds
+
+omit [OrderTopology k] [IsStrictOrderedRing k] in
+theorem Filter.Tendsto.asymptoticNhds_vadd_const {α} {f : α → V} {l : Filter α} {v : V} (p : P)
+    (hf : Tendsto f l (asymptoticNhds k V v)) :
+    Tendsto (fun x => f x +ᵥ p) l (asymptoticNhds k P v) := by
+  rw [← asymptoticNhds_vadd_pure, vadd_pure]
+  exact tendsto_map.comp hf
+
+omit [OrderTopology k] [IsStrictOrderedRing k] in
+theorem Filter.Tendsto.const_vadd_asymptoticNhds {α} {f : α → P} {l : Filter α} {v : V} (u : V)
+    (hf : Tendsto f l (asymptoticNhds k P v)) :
+    Tendsto (fun x => u +ᵥ f x) l (asymptoticNhds k P v) := by
+  rw [← vadd_asymptoticNhds u, ← Filter.map_vadd]
+  exact tendsto_map.comp hf
+
+theorem asymptoticNhds_smul (v : V) {c : k} (hc : 0 < c) :
+    asymptoticNhds k P (c • v) = asymptoticNhds k P v := by
```

Comment:
Should this be simp?

Reply:
simp doesn't care ;-) It will still use it if you have `0 < c` or if it can derive it itself